add new module with integration tests to manage gitlab project variables

markuman opened the pull request:
##### SUMMARY

A new module to manage gitlab project variables

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

gitlab_project_variable
<!--- Write the short name of the module, plugin, task or feature below -->

##### ADDITIONAL INFORMATION

The integration test can be run on any gitlab host. Just add three variable to `test/integration/integration_config.yml` and run `test/runner/ansible-test integration -v --python 3.7 gitlab_project_variable --docker --allow-unsupported`

```diff
diff --git a/test/integration/integration_config.yml b/test/integration/integration_config.yml
index 6b47a35dc1..60605ed585 100644
--- a/test/integration/integration_config.yml
+++ b/test/integration/integration_config.yml
@@ -1,3 +1,7 @@
 ---
 win_output_dir: 'C:\ansible_testing'
 output_dir: ~/ansible_testing
+gitlab_host: https://gitlab.com
+gitlab_login_token: secret-access-token
+gitlab_project_name: markuman/dotfiles
+
```


##### test.yml

```yml
---
- hosts: localhost
  gather_facts: False

  tasks:
    - name: Set or update some CI/CD variables
      gitlab_project_variable:
        api_url: https://gitlab.com
        api_token: secret-access-token
        project: markuman/dotfiles
        purge: True
        vars:
          ACCESS_KEY_ID: abc123
          SECRET_ACCESS_KEY: 321cba

```

##### invoke test.yml

```
markuman:~/git/ansible$ ansible-playbook test.yml 

PLAY [localhost] *******************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************

TASK [Set or update some CI/CD variables] ******************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************
changed: [localhost]

PLAY RECAP *************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

markuman:~/git/ansible$ ansible-playbook test.yml 

PLAY [localhost] *******************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************

TASK [Set or update some CI/CD variables] ******************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************
ok: [localhost]

PLAY RECAP *************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

Commit:
add new module with integration tests to manage gitlab project variables

Commit:
fix invalid yaml in DOCUMENTATION variable and don't import * from ansible module

Commit:
remove extends_documentation_fragment and put imports after DOCUMENTATION/EXAMPLES/RETURN/ANSIBLE_METADATA

Commit:
fix author in documentation and remove import from display

Commit:
add alias file for integration test

Commit:
split long lines and try to fix the author key

remove tailing whitespace

Commit:
replace email address with github username

Commit:
adding the at style to username

Commit:
add metaclass and future import

Commit:
add state variable to be able to delete selected variables

Commit:
add test with state = absent

Commit:
update documentation. scheme is necessary

Lunik commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,192 @@
+    default: present
+    type: str
+    choices: ["present", "absent"]
+  server_url:
```
Don't use `server_url` and `login_token` they will be deprecated on all gitlab modules in Ansible `2.10`.
Instead use parameters from fragment `auth_basic` api and `from ansible.module_utils.api import basic_auth_argument_spec`

Lunik commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,192 @@
+      - The path and name of the project
+    required: true
+    type: str
+  purge_vars:
```
Maybe `purge` with the description is enough to be understandable.

markuman replied:
https://github.com/markuman/ansible/commit/5fb79f16d2eb2bfc8a6c85a0cfa5ad9d0133bdea  
https://github.com/markuman/ansible/commit/e8767ea5ce9238443b620df7f827ae0dae52a89a

Lunik commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,192 @@
+
+EXAMPLES = '''
+- name: Set or update some CI/CD variables
+  gitlab_project_variables:
```
Your module is named `gitlab_project_variable` don't use plural in your example

markuman replied:
https://github.com/markuman/ansible/commit/ac165268041912b6c6eb423a85e44d9e5b2ce8f0

Lunik commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,192 @@
+      - ACCESS_KEY_ID: abc123
+'''
+
+RETURN = '''# '''
```
Please complete the `RETURN` section with the appropriate attributes

markuman replied:
https://github.com/markuman/ansible/commit/52fa05633dae34520e99081918f93e35879818e4

Lunik commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,192 @@
+def native_python_main(server_url, login_token, project_name, purge_vars, var_list, state):
+
+    change = False
+    this_gitlab = gitlab_project_variables(
```
Put the object creation out of this function in the `main` to be more readable.
Then only pass the gitlab object to your function.

markuman replied:
https://github.com/markuman/ansible/commit/d861dad27d61d89f8c992355f532d97dbcbc6188  
https://github.com/markuman/ansible/commit/ae0157ddc21bab5120c1e71a0cb8c50c686c7c82

Lunik commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,192 @@
+
+    def __init__(self, login_token, project_name, server_url):
+        self.repo = gitlab.Gitlab(
+            '{server_url}'.format(server_url=server_url), private_token=login_token)
```
What's the point of the `format` here ?

markuman replied:
https://github.com/markuman/ansible/commit/62ce2a5ac3f9d58a560e903ae8ec195c696e5c17

Lunik commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,192 @@
+    default: false
+    required: false
+    type: bool
+  vars:
```
Since project variables are uniques, a dictionary seems to be more appropriate than a list.

markuman replied:
https://github.com/markuman/ansible/commit/52fa05633dae34520e99081918f93e35879818e4  
https://github.com/markuman/ansible/commit/77f980f4486fc02781e44404933c31888a2dd239

Lunik commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,192 @@
+    name: markuman/dotfiles
+    purge_vars: False
+    vars:
+      - ACCESS_KEY_ID: abc123
```
Since each of your attributes have only one attributes, what's the point of making a list ?
Usage of a simple dictionary should be more easier to understand.
```yaml
vars:
  ACCESS_KEY_ID: abc123
  SECRET_ACCESS_KEY: 321cba
```

markuman replied:
https://github.com/markuman/ansible/commit/52fa05633dae34520e99081918f93e35879818e4  
https://github.com/markuman/ansible/commit/77f980f4486fc02781e44404933c31888a2dd239

Lunik commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,192 @@
+        )
+    )
+
+    if not HAS_GITLAB_PACKAGE:
```
Please follow best practice for this fail message
```python
if not HAS_GITLAB_PACKAGE:
    module.fail_json(msg=missing_required_lib("python-gitlab"), exception=GITLAB_IMP_ERR)
```

Lunik commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,192 @@
+    change = native_python_main(
+        server_url, login_token, project_name, purge_vars, var_list, state)
+
+    module.exit_json(changed=change, gitlab_project_variables=None)
```
Should be interesting to return data to the user, like variables `added`, `updated`, `removed

Lunik commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,192 @@
+      - Gitlab access token with api permissions.
+    required: true
+    type: str
+  name:
```
Change this parameter for a more explicite one : `project`
This one should handle project `path` or `id`

markuman replied:
https://github.com/markuman/ansible/commit/c2de76da386ebbc6770acca6dfdfe98bae57db87  
https://github.com/markuman/ansible/commit/ae0157ddc21bab5120c1e71a0cb8c50c686c7c82#diff-cde424ddaa0e494488bbf72147a63aa8R203  
https://github.com/markuman/ansible/commit/150309e9a7fdd8d03e3957c0a731fbb4c454545a

Lunik commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,192 @@
+
+from ansible.module_utils.basic import AnsibleModule, missing_required_lib
+
+try:
```
Use best practice to handler this error : 

```python
GITLAB_IMP_ERR = None
try:
    import gitlab
    HAS_GITLAB_PACKAGE = True
except Exception:
    GITLAB_IMP_ERR = traceback.format_exc()
    HAS_GITLAB_PACKAGE = False
```

markuman replied:
https://github.com/markuman/ansible/commit/f8a31db890dc4cb23724b7ca174c7de70dd84a74  
https://github.com/markuman/ansible/commit/ae0157ddc21bab5120c1e71a0cb8c50c686c7c82#diff-cde424ddaa0e494488bbf72147a63aa8L206

Lunik commented:
This module should be able to handle Ansible`check_mode` argument

Commit:
use singular in exmaple section

Commit:
use key purge instead of purge_vars

use purge instead of purge_vars also in the integration test

Commit:
create gitlab object in the ansible main function

Commit:
remove usedless .format

Commit:
follow best practice fail message

Commit:
add return documentation, return information about which variables were added, updated or removed and catch gitlab api auth error

Commit:
use module_utils.api with api_url and api_token

Commit:
use dict instead of list for vars

Commit:
use project name instead of name as playbook key

Commit:
add ansible checkmode_support, reduce variables in gitlab_project_variables class, remove wrong/duplicated HAS_GITLAB_PACKAGE check

markuman commented:
@Lunik Thanks a lot for your detailed review.  
I hope I took into account everything you've mentioned.

Lunik commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,245 @@
+    - "Markus Bergholz (@markuman)"
+requirements:
+  - python >= 2.7
+  - python-gitlab python module
```
Use doc fragment
```yaml
extends_documentation_fragment:
  - auth_basic
```

Lunik commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,245 @@
+    default: present
+    type: str
+    choices: ["present", "absent"]
+  api_url:
```
`api_url` and `api_token` shouldn't be referenced since there is the `extends_documentation_fragment` importing them automaticaly

Lunik commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,245 @@
+  type: str
+  sample: "Failed to connect to Gitlab server: 401: 401 Unauthorized"
+
+gitlab_project_variable:
```
I think that renaming the ouput variable `project_variable` is engouht

Lunik commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,245 @@
+    argument_spec = basic_auth_argument_spec()
+    argument_spec.pop('api_username')
+    argument_spec.pop('api_password')
+    argument_spec.pop('validate_certs')
```
Don't pop this parameters. They sould be accessible since they are imported with the documentation.

Lunik commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,245 @@
+
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        required_together=[
```
```python
mutually_exclusive=[
            ['api_username', 'api_token'],
            ['api_password', 'api_token'],
        ],
        required_together=[
            ['api_username', 'api_password'],
        ],
        required_one_of=[
            ['api_username', 'api_token']
        ],
        supports_check_mode=True,
```

Commit:
use extends_documentation_fragment and don't pop elements from basic_auth_argument_spec

Commit:
use just project_variable as output variable

Commit:
update mutually_exclusive as suggested

Commit:
re-add api_token documentation, because it is not included in api basic auth

markuman commented:
@Lunik Thx again. I've reworked the five suggestions

Lunik commented:
Great work, can you close reviews is you have properly made the modifications

Lunik commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,246 @@
+    def list_all_project_variables(self):
+        raw_variable_list = self.project.variables.list()
+        retval = []
+        if len(raw_variable_list) > 0:
```
The `if` statement is useless.
If the list is empty, then le `for` loop will not run.

Lunik commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,246 @@
+        retval = []
+        if len(raw_variable_list) > 0:
+            for item in raw_variable_list:
+                retval.append(item.get_id())
```
I think here, you should keep the whole `item`.
I will prevent unnecessary request to the API in then `update_variable` function : `self.project.variables.get(key)`

markuman replied:
I've found a bug :D I guess I introduced it while adding the check_mode for the update method().  
I added one more integration test for it.  
However, this will reduce the api requests of the gitlab inuse, but it needs some reworks of my "main" loop.

Lunik commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,246 @@
+def main():
+    argument_spec = basic_auth_argument_spec()
+    argument_spec.update(
+        api_token=dict(required=True, no_log=True, type='str'),
```
Simplify this by not adding required=False, and always starting with the type.
Since every parameter has a type, starting with this adds consistency.

markuman replied:
@Lunik but the api_token is required. Why should it be set to false?

Lunik replied:
On all your parameters, if there are not required, there is no need to specify the `required=False`.
But if they are, you should speficy `required=True`

Lunik replied:
`api_token` is a required parameter

Lunik commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,246 @@
+            change = True
+            return_value['removed'].append(item)
+
+    existing_variables = this_gitlab.list_all_project_variables()
```
It's seems that this statement is useless
```diff
- existing_variables = this_gitlab.list_all_project_variables()
```

Commit:
remove useless statement

remove unnecessary if

Commit:
add one test with a changing value

Commit:
put type at first position

Commit:
keep item to reduce api calls, build array and keep indexes by replacing with None instead of poping

Commit:
more asserts

markuman commented:
@Lunik Do you have some time to make a final review again?

waheedi commented:
@lunik @markuman i had some free time to go over the new module, it seems all is working as expected from a functional point of view, I have ran the tests and all were successful. 

I still have a question regarding the above test.yml, specifically this `gitlab_project_variables:` should not it be `gitlab_project_variable:` ?

It looks good overall 👍

markuman commented:
@waheedi Thanks for your review!

> I still have a question regarding the above test.yml, specifically this gitlab_project_variables: should not it be gitlab_project_variable: ?

That was just a typo.

Lunik commented:
shipit

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,246 @@
+  - When a project variable does not exists, it will be created.
+  - When a project variable does exists, its value will be updated when the values are different.
+  - Variables which are presented in the playbook, but are not presented in the Gitlab project,
+    they stay untouched (purged_vars = False) or will be deleted (purged_vars = True).
```
```suggestion
    they stay untouched (I(purged_vars) is C(false)) or will be deleted (I(purged_vars) is C(true)).
```
 - For booleans, use `true`/`false` or `yes`/`no` in documentation.
 - Use `I(...)` to mark options, and `C(...)` to mark values.

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,246 @@
+    choices: ["present", "absent"]
+  api_token:
+    description:
+      - Gitlab access token with api permissions.
```
```suggestion
      - Gitlab access token with API permissions.
```

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,246 @@
+
+DOCUMENTATION = '''
+module: gitlab_project_variable
+short_description: Creates/updates/deletes Gitlab Projects Variables
```
I think the official spelling is GitLab with a capital L.
```suggestion
short_description: Creates/updates/deletes GitLab Projects Variables
```

felixfontein replied:
There are more occurences of Gitlab without a captial L.

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,246 @@
+      ACCESS_KEY_ID: abc123
+      SECRET_ACCESS_KEY: 321cba
+
+- name: delete one variable
```
```suggestion
- name: Delete one variable
```
For consistency (compare the other example)

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,246 @@
+  sample: "Success"
+
+result:
+  description: json parsed response from the server
```
```suggestion
  description: JSON parsed response from the server
```

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,246 @@
+  type: dict
+
+error:
+  description: the error message returned by the Gitlab API
```
```suggestion
  description: The error message returned by the Gitlab API
```
The same for the next one.

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,246 @@
+  returned: always
+  type: dict
+
+error:
```
I don't think you ever return this.

markuman replied:
see https://github.com/ansible/ansible/pull/56574#discussion_r298044871

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,246 @@
+  type: str
+  sample: "Success"
+
+result:
```
You are not returning this.

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,246 @@
+'''
+
+RETURN = '''
+msg:
```
You don't return this one as well. Only specify things in `RETURN` which you actually return!

markuman replied:
@felixfontein  thx for your review.  

`msg` and `error` are used. But only in case of connection error.  
See line 226 and 228.  
But true, `result` was not used.

felixfontein replied:
`error` is not returned in either of them. The call to `fail_json` only returns `msg`.

markuman replied:
True. I've removed it.

markuman committed:
Update lib/ansible/modules/source_control/gitlab_project_variable.py

Co-Authored-By: Felix Fontein <felix@fontein.de>

Update lib/ansible/modules/source_control/gitlab_project_variable.py

Co-Authored-By: Felix Fontein <felix@fontein.de>

Update lib/ansible/modules/source_control/gitlab_project_variable.py

Co-Authored-By: Felix Fontein <felix@fontein.de>

Update lib/ansible/modules/source_control/gitlab_project_variable.py

Co-Authored-By: Felix Fontein <felix@fontein.de>

Update lib/ansible/modules/source_control/gitlab_project_variable.py

Co-Authored-By: Felix Fontein <felix@fontein.de>

Update lib/ansible/modules/source_control/gitlab_project_variable.py

Co-Authored-By: Felix Fontein <felix@fontein.de>

Commit:
remove unused return key from documentation

msg is only returned when failed

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,241 @@
+    type: bool
+  vars:
+    description:
+      - A list of key value pairs
```
```suggestion
      - A list of key value pairs.
```
Descriptions should be complete sentences.

markuman committed:
Update lib/ansible/modules/source_control/gitlab_project_variable.py

Co-Authored-By: Felix Fontein <felix@fontein.de>

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,241 @@
+  - When a project variable does not exists, it will be created.
+  - When a project variable does exists, its value will be updated when the values are different.
+  - Variables which are presented in the playbook, but are not presented in the Gitlab project,
+    they stay untouched (I(purged_vars) is C(false)) or will be deleted (I(purged_vars) is C(true)).
```
What is `purged_vars`, BTW? I now saw that this is not an option for this module.

markuman replied:
Ah. My origin version used `purged_vars`. But I changed it because @Lunik suggested to change it just to `purge`. See https://github.com/ansible/ansible/pull/56574#discussion_r285394389  
But I forget to change it in the documentation also.

Commit:
remove error key, because it is not returned

Commit:
change also documentation from purged_vars to purge

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,235 @@
+  sample: "Success"
+
+project_variable:
+  description: three lists of the variablenames which were added, updated or removed.
```
```suggestion
  description: Three lists of the variablenames which were added, updated or removed.
```
The code apparently returns four lists.

Could you document them as well? As an example of how to do that, take a look here: https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/crypto/acme/acme_account_info.py#L76-L101

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,235 @@
+  - When a project variable does not exists, it will be created.
+  - When a project variable does exists, its value will be updated when the values are different.
+  - Variables which are presented in the playbook, but are not presented in the Gitlab project,
+    they stay untouched (I(purged) is C(false)) or will be deleted (I(purged) is C(true)).
```
```suggestion
    they stay untouched (I(purge) is C(false)) or will be deleted (I(purge) is C(true)).
```

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,235 @@
+description:
+  - When a project variable does not exists, it will be created.
+  - When a project variable does exists, its value will be updated when the values are different.
+  - Variables which are presented in the playbook, but are not presented in the Gitlab project,
```
```suggestion
  - Variables which are present in the playbook, but are not present in the Gitlab project,
```

markuman committed:
Update lib/ansible/modules/source_control/gitlab_project_variable.py

Co-Authored-By: Felix Fontein <felix@fontein.de>

Update lib/ansible/modules/source_control/gitlab_project_variable.py

Co-Authored-By: Felix Fontein <felix@fontein.de>

Update test/integration/targets/gitlab_project_variable/tasks/main.yml

Co-Authored-By: Felix Fontein <felix@fontein.de>

Update test/integration/targets/gitlab_project_variable/tasks/main.yml

Co-Authored-By: Felix Fontein <felix@fontein.de>

pratikgadiya12 commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,235 @@
+module: gitlab_project_variable
+short_description: Creates/updates/deletes GitLab Projects Variables
+description:
+  - When a project variable does not exists, it will be created.
```
When a project variable does not exist, it will be created.

pratikgadiya12 commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,235 @@
+short_description: Creates/updates/deletes GitLab Projects Variables
+description:
+  - When a project variable does not exists, it will be created.
+  - When a project variable does exists, its value will be updated when the values are different.
```
exist

pratikgadiya12 commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,235 @@
+    they stay untouched (I(purge) is C(false)) or will be deleted (I(purge) is C(true)).
+version_added: "2.9"
+author:
+    - "Markus Bergholz (@markuman)"
```
indentation here seems different. I think there are extra spaces then rest indentation above

pratikgadiya12 commented:
Could you please add "check_mode" tests?

felixfontein commented on test/integration/targets/gitlab_project_variable/tasks/main.yml:
```diff
@@ -0,0 +1,185 @@
+- name: set two test variables state must be changed
+  assert:
+    that:
+      - gitlab_project_variable_state.changed
```
```suggestion
      - gitlab_project_variable_state is changed
```

felixfontein commented on test/integration/targets/gitlab_project_variable/tasks/main.yml:
```diff
@@ -0,0 +1,185 @@
+- name: re-set two test variables state must not be changed
+  assert:
+    that:
+      - not gitlab_project_variable_state.changed
```
```suggestion
      - gitlab_project_variable_state is not changed
```
(probably some more places for both)

felixfontein commented:
Check mode tests usually work as follows. Every test consists of four steps:
 1) do change in check mode
 2) do change for real
 3) do change again in check mode
 4) do change again for real

(3 and 4 can be swapped)

The idea is that both 1 and 2 should result in changes, and 3 and 4 not (idempotency).

Commit:
remove extra spaces

fix wrong spelling

Commit:
expand return value documentation with examples

Commit:
add check_mode test

reorder tests. first the check_mode test, later all other tests

markuman commented:
@felixfontein @pratikgadiya12
Thanks for your time and review.  
The check_mode test is now included.

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,256 @@
+      type: list
+      sample: "['ACCESS_KEY_ID', 'SECRET_ACCESS_KEY']"
+    updated:
+      description: A list of variables where their values has changed.
```
```suggestion
      description: A list of variables whose values were changed.
```

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,256 @@
+      type: list
+      sample: "['ACCESS_KEY_ID', 'SECRET_ACCESS_KEY']"
+    present:
+      description: A list of variables which exist.
```
Are these present before or after the module ran?

markuman replied:
They are present before the module ran.

felixfontein replied:
If `purge` is `true`, it will be empty according to the code you pasted below.

markuman replied:
Yes. `purge` is the point of view.  
If `purge` is `true`, `present` will be empty.  
If `purge` is `false`, `present` consists already existing variables.

felixfontein replied:
Above you said "They are present before the module ran". In that case, this list should not be empty if `purge` is `true`, but contain the list of variables which were there before the module ran.

felixfontein commented on test/integration/targets/gitlab_project_variable/tasks/main.yml:
```diff
@@ -0,0 +1,250 @@
+    that:
+      - gitlab_project_variable_state.changed
+      - gitlab_project_variable_state.project_variable.added|length == 0
+      - gitlab_project_variable_state.project_variable.present|length == 0
```
Shouldn't this be 1 since `ACCESS_KEY_ID` is still there (unchanged)? And `ACCESS_KEY_ID` already was there as well, so from the description, I would actually expect 2.

felixfontein replied:
I guess `present` means "of the variables specified in the module options, the ones which were already there and haven't changed". So `unchanged` would be a better term? Also mention in the description that this only considers variables listed in the options.

markuman replied:
> Shouldn't this be 1 since ACCESS_KEY_ID is still there (unchanged)? And ACCESS_KEY_ID already was there as well, so from the description, I would actually expect 2.

hmm yeah, interesting. That's wrong. I review it.

markuman replied:
@felixfontein 

I'm not sure what's the better/more readable variant to fix this issue

```python
    existing_variables = list(filter(None, existing_variables))
    if len(existing_variables) > 0 and purge:
        for item in existing_variables:
            this_gitlab.delete_variable(item)
            change = True
            return_value['removed'].append(item)
    elif len(existing_variables) > 0:
        for item in existing_variables:
            return_value['present'].append(item)
```

vs

```python
    existing_variables = list(filter(None, existing_variables))
    if len(existing_variables) > 0:
        for item in existing_variables:
            if purge:
                this_gitlab.delete_variable(item)
                change = True
                return_value['removed'].append(item)
            else:
                return_value['present'].append(item)
```

felixfontein replied:
How about
```.py
    existing_variables = list(filter(None, existing_variables))
    if purge:
        for item in existing_variables:
            this_gitlab.delete_variable(item)
            change = True
            return_value['removed'].append(item)
    else:
        return_value['present'].extend(existing_variables)
```

markuman replied:
>  So `unchanged` would be a better term?

I thought `present` is the default naming in ansible

felixfontein replied:
`present` is the default naming for a *state*. I would find it ok if this would be the list of variables which are there after the module ran. Or maybe even variables which were there before the module ran. But your definition is something else, so IMO the name should be different.

markuman committed:
Update lib/ansible/modules/source_control/gitlab_project_variable.py

Co-Authored-By: Felix Fontein <felix@fontein.de>

Commit:
fix existing keys in 'present' array

rework key handling (reduce code)

fix integration tests

use untouched instead of present to identify unchanged variable keys

fix wrong replacement

minor fixes on request

set aliases to unsupported, because the test succeed

remove posix group1 because it conflicts with unsupported

remove useless item from aliases

felixfontein commented:
BTW, it would help if you could squash *some* of the commits to reduce the number of commits to < 50.

Lunik commented:
All of your commits will be squash during the merge process. There is no need to squash them now.

markuman commented:
@Lunik @felixfontein @pratikgadiya12 do you have some time so that we can finish this?

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,262 @@
+description:
+  - When a project variable does not exist, it will be created.
+  - When a project variable does exist, its value will be updated when the values are different.
+  - Variables which are untouched in the playbook, but are not untouched in the Gitlab project,
```
```suggestion
  - Variables which are untouched in the playbook, but are not untouched in the GitLab project,
```
Please do a search and replace (with 'match case' enabled) to fix this in all cases.

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,262 @@
+options:
+  state:
+    description:
+      - create or delete project variable.
```
```suggestion
      - Create or delete project variable.
```

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,262 @@
+    type: str
+  project:
+    description:
+      - The path and name of the project
```
```suggestion
      - The path and name of the project.
```

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,262 @@
+    api_url: https://gitlab.com
+    api_token: secret_access_token
+    project: markuman/dotfiles
+    purge: False
```
```suggestion
    purge: false
```
Docs should use either `true`/`false` or `yes`/`no`, but not other boolean variants.

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,262 @@
+    description:
+      - When set to true, all variables which are not untoucheded in the task will be deleted.
+    default: false
+    required: false
```
```suggestion
```
Don't add `required: false`.

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,262 @@
+    description:
+      - A list of key value pairs.
+    default: {}
+    required: false
```
Same here.

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,262 @@
+    HAS_GITLAB_PACKAGE = False
+
+
+class gitlab_project_variables(object):
```
Class names are usually written in CamelCase to distinguish them from variable/attribute/method/function names.

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,262 @@
+'''
+
+RETURN = '''
+msg:
```
You are only using `msg` during module failures, not during success. So the description is wrong. Also, `msg` is one of the standard return values, so it should not show up in the module return value documentation. Simply remove the whole `msg` block.

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,262 @@
+    def update_variable(self, var, value):
+        if var.value == value:
+            return False
+        if self._module.check_mode and var.value != value:
```
```suggestion
        if self._module.check_mode:
```
You already checked for equality above.

felixfontein commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,262 @@
+        module.fail_json(msg="Failed to connect to Gitlab server: %s" % to_native(e))
+    except (gitlab.exceptions.GitlabHttpError) as e:
+        module.fail_json(msg="Failed to connect to Gitlab server: %s. \
+            Gitlab remove Session API now that private tokens are removed from user API endpoints since version 10.2" % to_native(e))
```
I don't understand this second line. What exactly do you want to say here?

markuman replied:
It catches an connection error when someone tries to use the old session api on a new/updated gitlab server.  

And I just cheated at the other gitlab modules to make it uniform across all gitlab modules.

felixfontein replied:
I find this error message very hard to understand. I don't have any experience with GitLab servers, though :)

If the other gitlab_* modules have the same text, we should probably keep it here as well. Changing it could be done in another PR once this is merged.

felixfontein reviewed (approved):
LGTM

(haven't tested it, though)

felixfontein commented on test/integration/targets/gitlab_project_variable/aliases:
```diff
@@ -0,0 +1,3 @@
+shippable/posix/group1
+gitlab/ci
+disabled
```
Ah, one final question: why are the tests disabled? If the tests can't be run via CI, this should be `unsupported`; `disabled` is for tests which fail and are temporarily disabled until they are fixed.

markuman replied:
It's the same here. I followed the other gitlab modules and they have the same content in `aliases`.  
I guess they are disabled because no one provided credentials for the CI to test? But as you said, they must be `unsupported. I change it for my module.  
Later I will run the integration tests of the other modules and when they succeed, I can make a pr to set them to `unsupported` too.

mattclay commented on test/integration/targets/gitlab_project_variable/aliases:
```diff
@@ -0,0 +1,3 @@
+shippable/posix/group1
```
Remove this alias since it conflicts with `unsupported`.

mattclay commented on test/integration/targets/gitlab_project_variable/aliases:
```diff
@@ -0,0 +1,3 @@
+shippable/posix/group1
+gitlab/ci
```
Remove this alias since it has no effect.

felixfontein commented:
@Lunik do you want to take another look?

Lunik commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,254 @@
+
+class GitlabProjectVariables(object):
+
+    def __init__(self, module):
```
GitLab authentication should not be inside this class.
It should be in the main function and then you init your class with the gitlab_instance object in parameter.
Look at other GitLab module

Lunik commented on lib/ansible/modules/source_control/gitlab_project_variable.py:
```diff
@@ -0,0 +1,254 @@
+        module.fail_json(msg=missing_required_lib("python-gitlab"), exception=GITLAB_IMP_ERR)
+
+    try:
+        this_gitlab = GitlabProjectVariables(module=module)
```
Authentication should be handled here and not inside of the class

Lunik reviewed (changes requested):
Review authentification workflow

Commit:
rework gitlab connection

Lunik commented:
shipit

felixfontein commented:
Do I assume correctly that at least one of you ran the tests with the latest changes (and they succeeded)? If yes, I can merge this.

markuman commented:
> Do I assume correctly that at least one of you ran the tests with the latest changes (and they succeeded)? If yes, I can merge this.

@felixfontein  Yes :) I run it before I commit and push the changes.  
Here is a screencast.

![peekHQY05Z](https://user-images.githubusercontent.com/3920157/62348958-a2f26c00-b4fe-11e9-943c-bbd47a183cc1.gif)

felixfontein commented:
Great! Just wanted to make sure that they work, since they don't run in CI :)

felixfontein commented:
@markuman thanks for contributing this module for Ansible 2.9!
@Lunik thanks for reviewing!
